rust/parser/typeql.pest

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+eof_query = { SOI ~ query ~ EOI }
+eof_queries = { SOI ~ query+ ~ EOI }
+eof_pattern = { SOI ~ pattern ~ EOI }
+eof_patterns = { SOI ~ patterns ~ EOI }
+eof_definables = { SOI ~ definables ~ EOI }
+eof_variable = { SOI ~ pattern_variable ~ EOI }
+eof_label = { SOI ~ label ~ EOI }
+eof_schema_rule = { SOI ~ schema_rule ~ EOI }
+
+// TYPEQL QUERY LANGUAGE ========================================================
+
+query = { query_define          | query_undefine
+        | query_insert          | query_update
+        | query_delete
+        | query_match_group_agg | query_match_group
+        | query_match_aggregate | query_match
+        }
+    query_define = { DEFINE ~ definables }
+    query_undefine = { UNDEFINE ~ definables }
+    query_insert = { (MATCH ~ patterns)? ~ INSERT ~ variable_things }
+    query_update = { query_delete ~ INSERT ~ variable_things }
+    query_delete = { MATCH ~ patterns ~ DELETE ~ variable_things }
+    query_match = { MATCH ~ patterns ~ modifiers }
+
+// MATCH QUERY ANSWER GROUP AND AGGREGATE FUNCTIONS ============================
+
+    query_match_aggregate = { query_match ~ match_aggregate }
+    query_match_group = { query_match ~ match_group }
+    query_match_group_agg = { query_match ~ match_group ~ match_aggregate }
+
+// MATCH QUERY MODIFIERS =======================================================
+
+modifiers = { (filter ~ ";")? ~ (sort ~ ";")? ~ (offset ~ ";")? ~ (limit ~ ";")? }
+
+filter = { GET ~ VAR_ ~ ( "," ~ VAR_ )* }
+sort = { SORT ~ var_order ~ ( "," ~ var_order )* }
+var_order = { VAR_ ~ ORDER_? }
+offset = { OFFSET ~ LONG_ }
+limit = { LIMIT ~ LONG_ }
+
+
+// GET AGGREGATE QUERY =========================================================
+//
+// An aggregate function is composed of 2 things=
```

Comment:
stray `=` at the end? maybe `:` is better

Reply:
That's a "find and replace" side effect 😅